### PR TITLE
Add support for coredumps which are stored using systemd-coredump

### PR DESCRIPTION
**Pull request by jonasbb**
From the `coredumpctl` man page:

> coredumpctl is a tool that can be used to retrieve and process
> core dumps and metadata which were saved by systemd-coredump(8).

`systemd-coredump` works by setting a piped core pattern. For example:

```bash
$ cat /proc/sys/kernel/core_pattern
|/usr/lib/systemd/systemd-coredump %P %u %g %s %t %c %h %e
```

This commit detects `systemd-coredump` in the pattern and then uses
`coredumpctl`, which is the CLI utility for reading the coredumps.
`coredumpctl` is instructed to write the coredump to a file. The
coredump is specified using the process ID and the basename of the
executable.

This commit also updates some doc strings to fix them and make them
more useful.

---

I kept the core in `native_corefile_pipe` for the `apport` case identical. I just changed the control flow to add another case.

I tried to keep `systemd_coredump_corefile` as simple as possible. The /dev/null files are requires, as otherwise `coredumpctl` would print quite a lot of status information about the coredump every time.

I did not add any test, because I do not know if and how that would be possible, as it would require an operating system with `systemd-coredump` configured. A docker container is not enough, as the `/proc` filesystem is mounted from the host.

This fixes both #1177 and #1178.

### CODE REVIEW

File: pwnlib/elf/corefile.py [hunk 1]
```diff
@@ -1309,25 +1340,36 @@ def native_corefile(self):
...
+                return filename
+
+            # Pretend core_pattern was just 'core', and see if we come up with anything
+            self.kernel_core_pattern = 'core'
+            return self.native_corefile_pattern()
+        elif 'systemd-coredump' in self.kernel_core_pattern:
+            log.debug("Found systemd-coredump in core_pattern")
+            return self.systemd_coredump_corefile()
+        else:
+            log.warn_once("Unsupported core_pattern: %r", self.kernel_core_pattern)
```
Comment (jonasbb) [inline pwnlib/elf/corefile.py:1391]: Switching to a `,` instead of the `%` fixes #1177.

File: pwnlib/elf/corefile.py [hunk 2]
```diff
@@ -1294,12 +1295,42 @@ def apport_read_crash_data(self):
...
+                [
+                    "coredumpctl",
+                    "dump",
+                    "--output=%s" % filename,
+                    # Filter coredump by pid and filename
+                    str(self.pid),
+                    self.basename,
+                ],
+                stdout=open(os.devnull, 'w'),
+                stderr=open(os.devnull, 'w'),
```
Comment (Arusekk) [inline pwnlib/elf/corefile.py]: I would suggest replacing the doubled `open()` call with stderr set to `STDOUT`
```suggestion
                stderr=subprocess.STDOUT,
```

**Review by Arusekk — CHANGES_REQUESTED** [review]
I think that the PR is great, have you tested it on your machine? I also submitted a review of changes that may be worth considering.

**Comment by jonasbb** [comment]
Hi @Arusekk, I included your feedback and rebased the PR to the current dev branch.

Yes I tested it locally, and it works fine.
```python
p = process(['./crash', payload])
c = p.corefile
# [*] Process './crash' stopped with exit code -11 (SIGSEGV) (pid 17058)                                               
# [x] Parsing corefile...
# [*] '/tmp/tmp/core.17058'
#     Arch:      i386-32-little
#     EIP:       0x80491d6
#     ESP:       0x6161616e
#     Exe:       '/tmp/tmp/crash' (0x8049000)
#     Fault:     0x6161616e
# [+] Parsing corefile...: Done
```

**Review by Arusekk — APPROVED** [review]

**Pull request merged**